### PR TITLE
refactor(agent&vscode): connect vscode extension to agent via lsp.

### PR DESCRIPTION
**Pull request by icycodes**
Changes:
- [TabbyAgent] Now builds the protocol typedefs, the executable LSP server for Node, and the LSP server webworker for the browser, no longer outputting js libs.
- The completion context collection implementation has been transferred from the VSCode extension to the TabbyAgent, includes fetching git context, extracting relative declarations, searching for recently changed code snippets.
- [VSCode] Reworked the structure for connecting to TabbyAgent via LSP.

### CODE REVIEW

File: clients/vscode/src/chat/ChatViewProvider.ts [hunk 1]
```diff
@@ -10,18 +11,25 @@ export class ChatViewProvider implements WebviewViewProvider {
...
       enableScripts: true,
       localResourceRoots: [extensionUri],
     };
-    webviewView.webview.html = await this._getWebviewContent();
+    console.log("resolveWebviewView", { config: this.config.server });
+    webviewView.webview.html = this._getWebviewContent(this.config.server);
+    this.config.on("updatedServerConfig", () => {
+      console.log("updatedServerConfig", { config: this.config.server });
+      webviewView.webview.html = this._getWebviewContent(this.config.server);
+    });
```
Comment (icycodes) [inline clients/vscode/src/chat/ChatViewProvider.ts:32]: @wwayne  I added a listener for changes ofserver config.

File: clients/vscode/src/chat/ChatViewProvider.ts [hunk 2]
```diff
@@ -51,11 +59,11 @@ export class ChatViewProvider implements WebviewViewProvider {
     });
   }
 
-  private async _getWebviewContent() {
-    const { server } = agent().getConfig();
+  private _getWebviewContent(server: ServerConfig) {
     return `
       <!DOCTYPE html>
       <html lang="en">
+        <!--hash: ${hashObject(server)}-->
```
Comment (icycodes) [inline clients/vscode/src/chat/ChatViewProvider.ts:66]: @wwayne  I added this hash for force refreshing when server config did change.

File: clients/vscode/src/chat/dom.d.ts [hunk 3]
```diff
@@ -0,0 +1,4 @@
+// We don't include DOM types in typescript building config
+
+// FIXME: This is required by `@quilted/threads` and `tabby-chat-panel`
+declare type HTMLIFrameElement = unknown;
```
Comment (icycodes) [inline clients/vscode/src/chat/dom.d.ts:4]: @wwayne  I remove DOM types dependencies in this refactor.  But `HTMLIFrameElement` is used by `@quilted/threads` and `tabby-chat-panel`.  Added this dummy declaration for tsc checks.

File: clients/vscode/src/chat/ChatViewProvider.ts (hunk 1 above)
Comment (wwayne) [reply clients/vscode/src/chat/ChatViewProvider.ts:30]: not sure if will affect other stuff like theme syncing, I will do more testing after merge

**Pull request merged**